### PR TITLE
[Fortran] Getters for meta-infos

### PR DESCRIPTION
**Pull request by chovyy**
This adds getter subroutines for meta-infos to the Fortran interface (`fs_get_serializer_metainfo`, `fs_get_field_metainfo` and `fs_get_savepoint_metainfo`) and the corresponding functions in the FortranWrapper.

It also adds add-subroutines for `INTEGER(KIND=C_LONG)` to the Fortran interface.

And I have changed the signature 

of 
`serialbox-c/Savepoint::serialboxSavepointGetMetainfo(serialboxSavepoint_t* savepoint)`
to
`serialbox-c/Savepoint::serialboxSavepointGetMetainfo(const serialboxSavepoint_t* savepoint)`

and of
`serialboxSerializerGetGlobalMetainfo(serialboxSerializer_t* serializer)`
to
`serialboxSerializerGetGlobalMetainfo(const serialboxSerializer_t* serializer)`.

I have created tests for all new functions and subroutines.

This PR is based on the branch from [PR  #139](https://github.com/eth-cscs/serialbox2/pull/139), so it contains all these changes too.

### CODE REVIEW

File: src/serialbox-c/FortranWrapper.cpp [hunk 1]
```diff
@@ -172,6 +176,13 @@ void serialboxFortranLoc(const void* basePtr, intptr_t* loc)
 	*loc = (long) basePtr;
 }
 
+void serialboxFortranSerializerGetFieldRank(const void* serializer, const char* name, int* rank) {
+
+	const Serializer* ser = toConstSerializer(static_cast<const serialboxSerializer_t*>(serializer));
+
+    *rank = ser->getFieldMetainfoImplOf(name).dims().size();
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.cpp]: Indentation looks weird here...

File: src/serialbox-c/FortranWrapper.cpp [hunk 2]
```diff
@@ -249,6 +269,23 @@ void serialboxFortranSerializerAddMetainfoString(void* serializer, const char* k
   }
 }
 
+#define SERIALBOX_FORTRAN_SERIALIZER_GET_METAINFO(name, CXXType)                                            \
+  void serialboxFortranSerializerGetMetainfo##name(const void* serializer, const char* key, CXXType* value) { \
+	const serialboxSerializer_t* ser = static_cast<const serialboxSerializer_t*>(serializer);               \
+	serialboxMetainfo_t* metainfo = serialboxSerializerGetGlobalMetainfo(ser);                              \
+    *value = serialboxMetainfoGet##name(metainfo, key);                                                     \
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.cpp]: Indent?

File: src/serialbox-c/FortranWrapper.cpp [hunk 3]
```diff
@@ -377,3 +454,19 @@ void serialboxFortranSavepointAddMetainfoString(void* savepoint, const char* key
     serialboxFatalError(e.what());
   }
 }
+
+#define SERIALBOX_FORTRAN_SAVEPOINT_GET_METAINFO(name, CXXType)                                             \
+  void serialboxFortranSavepointGetMetainfo##name(const void* savepoint, const char* key, CXXType* value) { \
+	const serialboxSavepoint_t* sp = static_cast<const serialboxSavepoint_t*>(savepoint);                   \
+	serialboxMetainfo_t* metainfo = serialboxSavepointGetMetainfo(sp);                                      \
+    *value = serialboxMetainfoGet##name(metainfo, key);                                                     \
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.cpp]: indent?

File: src/serialbox-c/FortranWrapper.h [hunk 4]
```diff
@@ -97,12 +103,32 @@ void serialboxFortranSerializerGetFieldHalos(const void* serializer, const char*
...
+ * \param value       Destination object for the value of the existing element
+ * @{
+ */
+void serialboxFortranSerializerGetMetainfoBoolean(const void* serializer, const char* key, int* value);
+void serialboxFortranSerializerGetMetainfoInt32(const void* serializer, const char* key, int* value);
+void serialboxFortranSerializerGetMetainfoInt64(const void* serializer, const char* key, long* value);
+void serialboxFortranSerializerGetMetainfoFloat32(const void* serializer, const char* key, float* value);
+void serialboxFortranSerializerGetMetainfoFloat64(const void* serializer, const char* key, double* value);
+void serialboxFortranSerializerGetMetainfoString(const void* serializer, const char* key,
+                                                 const char** value);
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.h]: Why value is on a second line just for this one?

File: src/serialbox-c/FortranWrapper.h [hunk 5]
```diff
@@ -158,6 +186,31 @@ void serialboxFortranSerializerAddFieldMetainfoString(void* serializer, const ch
...
+ * This function corresponds to `fs_get_field_metainfo_X`
+ *
+ * \param serializer  Serializer to use
+ * \param field       Name of the field
+ * \param key         Key of the new element
+ * \param value       Destination object for the value of the existing element
+ * @{
+ */
+void serialboxFortranSerializerGetFieldMetainfoBoolean(const void* serializer, const char* field,
+													   const char* key, int* value);
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.h]: indent problem ... and for the 5 next as well

File: src/serialbox-c/FortranWrapper.h [hunk 6]
```diff
@@ -174,12 +227,32 @@ void serialboxFortranSerializerAddFieldMetainfoString(void* serializer, const ch
...
+ * \param value       Destination object for the value of the existing element
+ * @{
+ */
+void serialboxFortranSavepointGetMetainfoBoolean(const void* savepoint, const char* key, int* value);
+void serialboxFortranSavepointGetMetainfoInt32(const void* savepoint, const char* key, int* value);
+void serialboxFortranSavepointGetMetainfoInt64(const void* savepoint, const char* key, long* value);
+void serialboxFortranSavepointGetMetainfoFloat32(const void* savepoint, const char* key, float* value);
+void serialboxFortranSavepointGetMetainfoFloat64(const void* savepoint, const char* key, double* value);
+void serialboxFortranSavepointGetMetainfoString(const void* savepoint, const char* key,
+                                                const char** value);
```
Comment (clementval) [inline src/serialbox-c/FortranWrapper.h]: same here ... why here only on a second line

**Comment by chovyy** [comment]
It seems, that my IDE has a problem with tabs and spaces. I will fix this. 

The second line for the ...String functions is copied from the Add functions. I was wondering, too. I guess, some line length limit was once the reason for this. I can change this. No problem.

**Comment by chovyy** [comment]
I forgot to say: I have left out the getters for strings in the Fortran interface because I have no idea how to implement it right.

**Comment by clementval** [comment]
@chovyy Ok, no worries. Can you just put a comment with `! TODO ` and the getters that are not implemented?

**Comment by chovyy** [comment]
Done.

**Comment by chovyy** [comment]
Hi, I have changed the (Fortran) tests for the meta-info in that way that the serializer is destroyed and recreated before calling the meta-info getters and it turned out that the meta-infos are not actually written to the file. Does anyone has an explanation for that?

**Comment by thfabian** [comment]
The meta-data is buffered internally and only flushed to disk before you serialize a field. However, the serializer has a method [updateMetaData](https://eth-cscs.github.io/serialbox2/_doxygen/html/classserialbox_1_1SerializerImpl.html#a1250ec8ca1febff26bc20346a9c99577) which forces writing the meta-data to disk. You may want to wrap this function in the C interface.

**Comment by chovyy** [comment]
Wouldn't it make sense to call that method in the Serializer's destructor?

**Comment by chovyy** [comment]
Okay, I'm trying to understand it. 

The protocol of C++ (and C?) interface is: register field, add meta-info, write field. Correct? So, as long as you follow this protocol, you don't have a problem. 

But in the Fortran interface registration is embedded in the write routines. That means, you cannot add meta-data before writing, because then the field doesn't yet exist. But if you add meta-data after writing, the last written field will always come without meta-data. That means, for Fortran we definitely have to make an extra call the update method.

So, do you think, we should call that update method only from the Fortran destroy routine or generally in the Serializer's destructor to be more flexible?

**Comment by thfabian** [comment]
> Wouldn't it make sense to call that method in the Serializer's destructor?

`updateMetaData` can throw an exception which would result in termination of the app, see [here](https://stackoverflow.com/questions/130117/throwing-exceptions-out-of-a-destructor). You could, however, add this functionality to the C interface and call it in `serialboxSerializerDestroy`.

> The protocol of the C++ (and C?) interface is: register field, add meta-info, write field. Correct? So, as long as you follow this protocol, you don't have a problem.

Correct.

> So, do you think, we should call that update method only from the Fortran destroy routine or generally in the Serializer's destructor to be more flexible?

`serialboxSerializerDestroy` might be the place to do this.

**Comment by chovyy** [comment]
> updateMetaData can throw an exception which would result in termination of the app, see here. 
Okay, sounds convincing. 

> You could, however, add this functionality to the C interface and call it in serialboxSerializerDestroy.
Alright, that means we'll keep the C++ protocol as it is, but will have a final update in the C and Fortran interface.

**Pull request merged**